src/TParsec/Combinators.idr

```diff
@@ -31,33 +38,88 @@ guardM : (Alternative mn, Monad mn) =>
                                         choiceMap pure (Success.guardM f s)
 
+||| Constrains a parsed to succeed only when a predicate holds.
+|||
+||| Given a predicate on a value, and a parser of such value, return a parser 
+||| that will fail when the predicate is false and will return the value if
+||| the predicate is true.
+|||
+||| Unindexed signature: `(a -> Bool) -> Parser a -> Parser a`
 guard : (Alternative mn, Monad mn) =>
         (a -> Bool) -> All (Parser mn p a :-> Parser mn p a)
 guard p = guardM $ \v => toMaybe (p v) v
 
+||| Processes a token into a `Maybe` value.
+|||
+||| Given a function that map a parsed token into a Maybe a, this function
+||| returns a parser that fails when the token is mapped to `Nothing` and
+||| succeeds when the value is processed into a `Just`. The successful
+||| value is then unwrapped from the `Maybe` and `Parser a` is returned.
+|||
+||| Unindexed signature: `(Tok p -> Maybe a) -> Parser a`
 maybeTok : (Alternative mn, Monad mn, Inspect (Toks p) (Tok p)) =>
            (Tok p -> Maybe a) -> All (Parser mn p a)
 maybeTok p = guardM p anyTok
 
+||| Given a function (a -> b), transforms a `Parser a` into a `Parser b`.
+|||
+||| Map lifts a function from `a -> b` to `Parser a -> Parser b`. This 
+||| function signature does not follow the traditional `Functor` signature 
+||| (which is `(a -> b) -> F a -> F b`) due to the indexing rules that ensure 
+||| totality.
+|||
+||| Unindexed signature: `(a -> b) -> Parser a -> Parser b`
 map : Functor mn =>
       (a -> b) -> All (Parser mn p a :-> Parser mn p b)
 map f p = MkParser $ \le, ts => Functor.map (Success.map f) (runParser p le ts)
 
+||| Assuming the parser is successful, return the given value.
+|||
+||| Given a value `v` of type `a` this function will replace the parsed value
+||| by `v`, assuming it is successful. The parsed value is discarded.
+|||
+||| Unindexed signature: `b -> Parser a -> Parser b`
 cmap : Functor mn => b -> All (Parser mn p a :-> Parser mn p b)
 cmap b = map (\_ => b)
 
+||| A parser that always fail.
+|||
+||| Unindexed signature: `Parser a`
 fail : Alternative mn => All (Parser mn p a)
 fail = MkParser $ \_, _ => empty
 
+||| Given two parser, take the first one that succeeds.
+|||
+||| If the first parser fails, the second one will be attempted, if the second
+||| one fails the whole parser fails. This is anologous to an `or` operation.
+|||
+||| Unindexed signature: `Parser a -> Parser a -> Parser a`
 alt : Alternative mn =>
       All (Parser mn p a :-> Parser mn p a :->
            Parser mn p a)
 alt p q = MkParser $ \mlen, ts => runParser p mlen ts <|> runParser q mlen ts
 
+||| Given a list of parsers, use take the first one that succeeds, in order.
+|||
+||| Attempt parsing using each parser in the list. If all fail, then this 
+||| parser will fail too. If one succeeds, then the following parsers will
+||| not be attempted and the parser will succeed.
+|||
+||| Unindexed signature: `[Parser a] -> Parser a`
```

Comment:
Should probably say `List (Parser a)` to avoid haskellisms

Reply:
Oops, my ~~tongue~~ keystrokes slipped